src/plots/cartesian/align_period.js

```diff
@@ -0,0 +1,92 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+var Lib = require('../../lib');
+var ms2DateTime = Lib.ms2DateTime;
+var dateTime2ms = Lib.dateTime2ms;
+var incrementMonth = Lib.incrementMonth;
+var ONEDAY = require('../../constants/numerical').ONEDAY;
+
+module.exports = function alignPeriod(trace, ax, axLetter, vals) {
+    if(ax.type !== 'date') return vals;
+
+    var alignment = trace[axLetter + 'periodalignment'];
+    if(!alignment) return vals;
+
+    var period = trace[axLetter + 'period'];
+    var mPeriod, dPeriod;
+    if(isNumeric(period)) {
+        dPeriod = +period;
+        dPeriod /= ONEDAY; // convert milliseconds to days
+        dPeriod = Math.round(dPeriod);
+        if(dPeriod <= 0) return vals;
+    } else if(typeof period === 'string' && period.charAt(0) === 'M') {
+        var n = +(period.substring(1));
+        if(n > 0 && Math.round(n) === n) {
+            mPeriod = n;
+        } else return vals;
+    }
+
+    var calendar = ax.calendar;
+
+    var isStart = 'start' === alignment;
+    // var isMiddle = 'middle' === alignment;
+    var isEnd = 'end' === alignment;
+
+    var period0 = trace[axLetter + 'period0'];
+    var base = dateTime2ms(period0, calendar) || 0;
+
+    var newVals = [];
+    var len = vals.length;
+    for(var i = 0; i < len; i++) {
+        var v = vals[i] - base;
+
+        var dateStr = ms2DateTime(v, 0, calendar);
+        var d = new Date(dateStr);
+        var year = d.getUTCFullYear();
+        var month = d.getUTCMonth();
+        var day = d.getUTCDate();
```

Comment:
I think this approach is still going to have problems with time zones - specifically, I think the day will be wrong this way if you're in the eastern hemisphere.

There would be ways around this, but we've already solved these problems and all sorts of other edge cases with functions in `lib/dates` so you never directly need `Date` objects again and all your processing is with millisecond numbers. What I was thinking of in https://github.com/plotly/plotly.js/pull/5074#discussion_r490279882 was something like:
```js
var v = vals[i];
// guess at how many periods away from base we are
var nEstimated = Math.round((v - base) / (mPeriod * ONEAVGMONTH));
var endTime = incrementMonth(base, mPeriod * nEstimated, calendar);

// iterate to get the exact bounds before and after v
// there may be ways to make this faster, but most of the time
// you'll only execute each loop zero or one time.
while (endTime > v) {
    endTime = incrementMonth(endTime, -mPeriod, calendar);
}
while (endTime <= v) {
    endTime = incrementMonth(endTime, mPeriod, calendar);
}

// now we know endTime is the boundary immediately after v
// so startTime is obtained by incrementing backward one period.
var startTime = incrementMonth(endTime, -mPeriod, calendar);

newVals[i] = (
    isStart ? startTime :
    isEnd ? endTime :
    (startTime + endTime) / 2;
)
```

Reply:
Thanks for your review. 
Revised in https://github.com/plotly/plotly.js/pull/5074/commits/83b22a06105de15e9825973e76cf8a29ffc62048.